### ISSUE TITLE
Proposal: warn on unnecessary async wrapper in expect(...).rejects or expect(...).resolves

### ISSUE THREAD

**Issue by y-hsgw** [jest-community/eslint-plugin-jest#1722]
### Problem

It is common to write assertions for rejected promises using `.rejects`. However, sometimes developers write it in a more verbose and unnecessary way like this:

```ts
await expect(async () => {
  await somePromise();
}).rejects.toThrow('Error');
```

This can be simplified to:

```ts
await expect(somePromise()).rejects.toThrow('Error');
```

The async function wrapper adds no value and makes the test harder to read, especially when the only purpose of the function is to wrap a single `await` call.

### Proposal

I'd like to propose a lint rule that detects and warns on unnecessary async function wrappers passed to expect(...).rejects or expect(...).resolves.

There are two possible implementation paths:

#### Option 1: Extend `prefer-expect-resolves`

This rule already encourages usage of `.resolves` / `.rejects` over `try/catch` or other patterns.  
It may be natural to include this redundant async wrapper pattern as part of that rule’s responsibility.

#### Option 2: Introduce a new rule

Create a new rule such as `no-unnecessary-async-wrapper-in-expect` that focuses specifically on this anti-pattern.  
It would help keep test code concise and idiomatic.

### Example Code

**Invalid:**

```ts
await expect(async () => {
  await doSomethingAsync();
}).rejects.toThrow();
```

**Valid:**

```ts
await expect(doSomethingAsync()).rejects.toThrow();
```

### Discussion

I'm unsure whether this case is better suited as an extension of prefer-expect-resolves, or whether it's better introduced as a standalone rule. I’d love to hear your thoughts on that — or if there's another preferable approach, I'd really appreciate your guidance.

Thanks for your time!

**Comment by G-Rath** [comment]
While I don't think it'd be that hard to support, I'm not sure if it'd be that useful since this hasn't been raised before and I personally have never seen this being done - if you had any metrics that could be shared, that'd be interesting especially if it's about public repos that are doing this.

fwiw also I believe you can catch this with `no-restricted-syntax` with [`CallExpression[callee.name=expect][arguments.0.async=true]:matches([arguments.0.body.body.length=1], [arguments.0.body.type=CallExpression])`](https://explorer.eslint.org/#eslint-explorer=IntcInN0YXRlXCI6e1widG9vbFwiOlwiYXN0XCIsXCJjb2RlXCI6e1wiamF2YXNjcmlwdFwiOlwiYXdhaXQgZXhwZWN0KGFzeW5jICgpID0%2BIHtcXG4gIGF3YWl0IGRvU29tZXRoaW5nQXN5bmMoKTtcXG59KS5yZWplY3RzLnRvVGhyb3coKTtcXG5cXG5hd2FpdCBleHBlY3QoYXN5bmMgKCkgPT4ge1xcbiAgYXdhaXQgZG9Tb21ldGhpbmdBc3luYygpO1xcbiAgYXdhaXQgZG9Tb21ldGhpbmdBc3luYygpO1xcbn0pLnJlamVjdHMudG9UaHJvdygpO1xcblxcbmF3YWl0IGV4cGVjdChhc3luYyBmdW5jdGlvbigpIHtcXG4gIGF3YWl0IGRvU29tZXRoaW5nQXN5bmMoKTtcXG59KS5yZWplY3RzLnRvVGhyb3coKTtcXG5cXG5hd2FpdCBleHBlY3QoYXN5bmMgZnVuY3Rpb24oKSB7XFxuICBhd2FpdCBkb1NvbWV0aGluZ0FzeW5jKCk7XFxuICBhd2FpdCBkb1NvbWV0aGluZ0FzeW5jKCk7XFxufSkucmVqZWN0cy50b1Rocm93KCk7XFxuXFxuYXdhaXQgZXhwZWN0KGFzeW5jICgpID0%2BIGRvU29tZXRoaW5nQXN5bmMoKSkucmVqZWN0cy50b1Rocm93KCk7XCIsXCJqc29uXCI6XCIvKipcXG4gKiBUeXBlIG9yIHBhc3RlIHNvbWUgSlNPTiBoZXJlIHRvIGxlYXJuIG1vcmUgYWJvdXRcXG4gKiB0aGUgc3RhdGljIGFuYWx5c2lzIHRoYXQgRVNMaW50IGNhbiBkbyBmb3IgeW91LlxcbiAqXFxuICogVGhlIHRhYnMgYXJlOlxcbiAqXFxuICogLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxcbiAqICAgYmUgdXNlZnVsIHRvIHVuZGVyc3RhbmQgdGhlIHN0cnVjdHVyZSBvZiB0aGUgY29kZS4gWW91XFxuICogICBjYW4gdmlldyB0aGlzIHN0cnVjdHVyZSBhcyBKU09OIG9yIGluIGEgdHJlZSBmb3JtYXQuXFxuICpcXG4gKiBZb3UgY2FuIGNoYW5nZSB0aGUgd2F5IHRoYXQgdGhlIEpTT04gY29kZSBpcyBpbnRlcnByZXRlZFxcbiAqIGJ5IGNsaWNraW5nIFxcXCJKU09OXFxcIiBpbiB0aGUgaGVhZGVyIGFuZCBzZWxlY3RpbmcgZGlmZmVyZW50XFxuICogb3B0aW9ucy5cXG4gKlxcbiAqIFRoaXMgZXhhbXBsZSBpcyBpbiBKU09OQyBtb2RlLCB3aGljaCBhbGxvd3MgY29tbWVudHMuXFxuICovXFxuXFxue1xcbiAgICBcXFwia2V5MVxcXCI6IFt0cnVlLCBmYWxzZSwgbnVsbF0sXFxuICAgIFxcXCJrZXkyXFxcIjoge1xcbiAgICAgICAgXFxcImtleTNcXFwiOiBbMSwgMiwgXFxcIjNcXFwiLCAxZTEwLCAxZS0zXVxcbiAgICB9XFxufVwiLFwibWFya2Rvd25cIjpcIjwhLS1cXG5UeXBlIG9yIHBhc3RlIHNvbWUgTWFya2Rvd24gaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxudGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG5cXG5UaGUgdGFicyBhcmU6XFxuXFxuLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxcbmJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbmNhbiB2aWV3IHRoaXMgc3RydWN0dXJlIGFzIEpTT04gb3IgaW4gYSB0cmVlIGZvcm1hdC5cXG5cXG5Zb3UgY2FuIGNoYW5nZSB0aGUgd2F5IHRoYXQgdGhlIE1hcmtkb3duIGNvZGUgaXMgaW50ZXJwcmV0ZWRcXG5ieSBjbGlja2luZyBcXFwiTWFya2Rvd25cXFwiIGluIHRoZSBoZWFkZXIgYW5kIHNlbGVjdGluZyBkaWZmZXJlbnRcXG5vcHRpb25zLlxcblxcblRoaXMgZXhhbXBsZSBpcyBpbiBDb21tb25NYXJrIG1vZGUuXFxuLS0%2BXFxuXFxuIyBFU0xpbnQgTWFya2Rvd24gRXhhbXBsZVxcblxcblRoaXMgaXMgYW4gZXhhbXBsZSBvZiBhIE1hcmtkb3duIGZpbGUgdGhhdCBjYW4gYmUgcGFyc2VkXFxuYnkgRVNMaW50LiBNYXJrZG93biBpcyBhIHNpbXBsZSBtYXJrdXAgbGFuZ3VhZ2UgdGhhdCBpc1xcbm9mdGVuIHVzZWQgZm9yIGRvY3VtZW50YXRpb24uXFxuXFxuIyMgRmVhdHVyZXNcXG5cXG4tIE1ha2UgdGhpbmdzICppdGFsaWMqLCAqKmJvbGQqKiwgb3IgYGNvZGVgXFxuLSBDcmVhdGUgW2xpbmtzXShodHRwczovL2VzbGludC5vcmcpXFxuLSBTdXBwb3J0cyBIVE1MIDxzcGFuIHN0eWxlPVxcXCJjb2xvcjogcmVkO1xcXCI%2BZWxlbWVudHM8L3NwYW4%2BXFxuLSBMaXN0c1xcbiAgLSBOZXN0ZWQgbGlzdHNcIixcImNzc1wiOlwiLyoqXFxuICogVHlwZSBvciBwYXN0ZSBzb21lIENTUyBoZXJlIHRvIGxlYXJuIG1vcmUgYWJvdXRcXG4gKiB0aGUgc3RhdGljIGFuYWx5c2lzIHRoYXQgRVNMaW50IGNhbiBkbyBmb3IgeW91LlxcbiAqXFxuICogVGhlIHRhYnMgYXJlOlxcbiAqXFxuICogLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxcbiAqICAgYmUgdXNlZnVsIHRvIHVuZGVyc3RhbmQgdGhlIHN0cnVjdHVyZSBvZiB0aGUgY29kZS4gWW91XFxuICogICBjYW4gdmlldyB0aGlzIHN0cnVjdHVyZSBhcyBKU09OIG9yIGluIGEgdHJlZSBmb3JtYXQuXFxuICpcXG4gKiBZb3UgY2FuIGNoYW5nZSB0aGUgd2F5IHRoYXQgdGhlIENTUyBjb2RlIGlzIGludGVycHJldGVkXFxuICogYnkgY2xpY2tpbmcgXFxcIkNTU1xcXCIgaW4gdGhlIGhlYWRlciBhbmQgc2VsZWN0aW5nIGRpZmZlcmVudFxcbiAqIG9wdGlvbnMuXFxuICovXFxuXFxuQGltcG9ydCB1cmwoJ2h0dHBzOi8vZm9udHMuZ29vZ2xlYXBpcy5jb20vY3NzMj9mYW1pbHk9Um9ib3RvOndnaHRANDAwOzcwMCZkaXNwbGF5PXN3YXAnKTtcXG5cXG5ib2R5IHtcXG5cXHRmb250LWZhbWlseTogc2Fucy1zZXJpZjtcXG59XFxuXFxuaDEge1xcblxcdGNvbG9yOiAjMzMzO1xcbn1cXG5cXG5wIHtcXG5cXHRtYXJnaW46IDFlbSAwO1xcbn1cIixcImh0bWxcIjpcIjwhRE9DVFlQRSBodG1sPlxcbjwhLS1cXG5UeXBlIG9yIHBhc3RlIHNvbWUgSFRNTCBoZXJlIHRvIGxlYXJuIG1vcmUgYWJvdXRcXG50aGUgc3RhdGljIGFuYWx5c2lzIHRoYXQgRVNMaW50IGNhbiBkbyBmb3IgeW91LlxcblxcblRoZSB0YWJzIGFyZTpcXG5cXG4tIEFTVCAtIFRoZSBBYnN0cmFjdCBTeW50YXggVHJlZSBvZiB0aGUgY29kZSwgd2hpY2ggY2FuXFxuYmUgdXNlZnVsIHRvIHVuZGVyc3RhbmQgdGhlIHN0cnVjdHVyZSBvZiB0aGUgY29kZS4gWW91XFxuY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0Llxcbi0tPlxcblxcbjxodG1sIGxhbmc9XFxcImVuXFxcIj5cXG4gICAgPGhlYWQ%2BXFxuICAgICAgICA8bWV0YSBjaGFyc2V0PVxcXCJVVEYtOFxcXCI%2BXFxuICAgICAgICA8dGl0bGU%2BSFRNTDwvdGl0bGU%2BXFxuICAgIDwvaGVhZD5cXG4gICAgPGJvZHk%2BXFxuICAgICAgICA8cD5UZXh0PC9wPlxcbiAgICA8L2JvZHk%2BXFxuPC9odG1sPlwifSxcImxhbmd1YWdlXCI6XCJqYXZhc2NyaXB0XCIsXCJqc09wdGlvbnNcIjp7XCJwYXJzZXJcIjpcImVzcHJlZVwiLFwic291cmNlVHlwZVwiOlwibW9kdWxlXCIsXCJlc1ZlcnNpb25cIjpcImxhdGVzdFwiLFwiaXNKU1hcIjp0cnVlfSxcImpzb25PcHRpb25zXCI6e1wianNvbk1vZGVcIjpcImpzb25jXCIsXCJhbGxvd1RyYWlsaW5nQ29tbWFzXCI6ZmFsc2V9LFwiY3NzT3B0aW9uc1wiOntcImNzc01vZGVcIjpcImNzc1wiLFwidG9sZXJhbnRcIjpmYWxzZX0sXCJtYXJrZG93bk9wdGlvbnNcIjp7XCJtYXJrZG93bk1vZGVcIjpcImNvbW1vbm1hcmtcIixcIm1hcmtkb3duRnJvbnRtYXR0ZXJcIjpcIm9mZlwifSxcIndyYXBcIjp0cnVlLFwidmlld01vZGVzXCI6e1wiYXN0Vmlld1wiOlwidHJlZVwiLFwic2NvcGVWaWV3XCI6XCJmbGF0XCIsXCJwYXRoVmlld1wiOlwiZ3JhcGhcIn0sXCJwYXRoSW5kZXhcIjp7XCJpbmRleFwiOjAsXCJpbmRleGVzXCI6MX0sXCJlc3F1ZXJ5U2VsZWN0b3JcIjp7XCJzZWxlY3RvclwiOlwiQ2FsbEV4cHJlc3Npb25bY2FsbGVlLm5hbWU9ZXhwZWN0XVthcmd1bWVudHMuMC5hc3luYz10cnVlXTptYXRjaGVzKFthcmd1bWVudHMuMC5ib2R5LmJvZHkubGVuZ3RoPTFdLCBbYXJndW1lbnRzLjAuYm9keS50eXBlPUNhbGxFeHByZXNzaW9uXSlcIn19LFwidmVyc2lvblwiOjB9Ig%3D%3D)

**Comment by y-hsgw** [comment]
I didn't have any specific metrics; I just happened to notice this pattern being used more often recently, which made me curious—so I created the issue somewhat casually. 😄
That said, when I searched using the following query, I did come across this pattern here and there.
https://github.com/search?q=rejects.toThrow+language%3ATypeScript&type=code

> fwiw also I believe you can catch this with no-restricted-syntax

I see, so it can be detected that way too. Thanks for the reference!

**Comment by y-hsgw** [comment]
Thanks for your consideration!
I’m also on board with adding a new rule — that sounds like a solid direction.
